RestartManager.js

```diff
@@ -26,6 +28,12 @@ const RestartManager = (() => {
 
     function restartApp(onlyIfUpdateIsPending = false) {
         if (_allowed) {
+            if (_inProgress) {
+                log("A restart request is already in progress or queued");
+                return;
+            }
+            // The restart won't execute if `onlyIfUpdateIsPending === true` and there is no pending update.
+            _inProgress = !onlyIfUpdateIsPending || !!(NativeCodePush.getUpdateMetadata(CodePush.UpdateState.PENDING));
```

Comment:
Couldn't you set `__inProgress` to `true` and then set it to `false` when a `Promise` returned by `NativeCodePush.restartApp` resolves? That way, you set the flag to `true` immediately, which blocks subsequent calls to `restartApp` in the same flush cycle of the JS->native call queue. If the app actually restarts, then it doesn't matter, but at least you can guarantee that calls to `restartApp` can succeed after the native call which doesn't result in a restart has completed, and returned back to JS.